File: packages/octospaces-sdk/src/spaces/object-index.ts
/**
 * Headless reads + create-time seeding of a space's unified OBJECT INDEX.
 *
 * Both PRIVATE (encrypted) and PUBLIC (plaintext) spaces store their room/category
 * tree in `spaces/{spaceId}/objects/_index`. The `encryptor` parameter is null for
 * public spaces — the helpers pass data through unchanged.
 */
import { ConflictError } from '@drakkar.software/starfish-client';
import type { Encryptor, StarfishClient } from '@drakkar.software/starfish-client';

import type { ObjectNode, Room, SpaceVisibility } from '../core/types.js';
import type { Session } from '../sync/identity.js';
import { DEFAULT_CATEGORY, objectsToRoomCategories, seedIndexNodes } from '../objects/objects.js';
import type { SeedRoom } from '../objects/objects.js';
import { objIndexPull, objIndexPush } from '../sync/paths.js';
import { buildSpaceAccess, getSpaceAccess } from '../sync/space-access.js';

export type { SeedRoom };

function indexNodes(plain: Record<string, unknown>): ObjectNode[] {
  return Array.isArray((plain as { objects?: unknown }).objects)
    ? (plain as { objects: ObjectNode[] }).objects
    : [];
}

/**
 * Pull + (private: decrypt) + project a space's object index into the legacy
 * `{ rooms, categories }` shape. `encryptor` is null for a PUBLIC space.
 * Returns null on failure or an empty index.
 */
export async function readIndexRooms(
  client: StarfishClient,
  encryptor: Encryptor | null,
  indexPath: string,
  spaceId: string,
): Promise<{ rooms: Room[]; categories: string[] } | null> {
  try {
    const res = await client.pull(indexPath).catch(() => null);
    if (!res?.data) return null;
    const plain = encryptor
      ? await encryptor.decrypt(res.data as Record<string, unknown>)
      : (res.data as Record<string, unknown>);
    const cats = objectsToRoomCategories(indexNodes(plain), spaceId, DEFAULT_CATEGORY);
    if (!cats) return null;
    return { rooms: cats.flatMap((c) => c.rooms), categories: cats.map((c) => c.name) };
  } catch {
    return null;
  }
}

/**
 * Read a space's index rooms + categories, resolving access automatically.
 * Pass `reg` (from `readRooms`) so the accessor picks the right auth mode.
 */
export async function readSpaceIndexRooms(
  session: Session,
  spaceId: string,
  reg: { owner: string | null; members: string[]; visibility?: SpaceVisibility },
): Promise<{ rooms: Room[]; categories: string[] } | null> {
  if (reg.owner === null && reg.visibility !== 'public') return null;
  try {
    const { encryptor, client } = await getSpaceAccess(spaceId, session, reg);
    return await readIndexRooms(client, encryptor, objIndexPull(spaceId), spaceId);
  } catch {
    return null;
  }
}

/** SOFT read — never throws, never mints. Returns an empty array on any failure. */
export async function readSpaceRooms(
  session: Session,
  spaceId: string,
  hint?: { visibility?: SpaceVisibility },
): Promise<Room[]> {
  const access = await buildSpaceAccess(session, spaceId, hint).catch(() => null);
  if (!access) return [];
  const idx = await readIndexRooms(access.client, access.encryptor, objIndexPull(spaceId), spaceId);
  return idx?.rooms ?? [];
}

/**
 * Write the create-time seed into a space's index doc with an already-open access handle.
 * Accepts a nullable encryptor — plaintext push when null (public spaces).
 * Idempotent: a no-op if the index doc already exists (either encrypted or plaintext).
 */
export async function pushIndexSeed(
  client: StarfishClient,
  encryptor: Encryptor | null,
  spaceId: string,
  rooms: SeedRoom[],
): Promise<void> {
  const res = await client.pull(objIndexPull(spaceId)).catch(() => null);
  const existing = res?.data as Record<string, unknown> | undefined;
  if (existing?._encrypted || Array.isArray(existing?.objects)) return;
  const nodes = seedIndexNodes(rooms, Date.now());
  const payload = encryptor
    ? await encryptor.encrypt({ objects: nodes }) as Record<string, unknown>
    : { objects: nodes };
  await client.push(objIndexPush(spaceId), payload, res?.hash ?? null);
}

/**
 * Seed a brand-new space's index as the OWNER.
 * For private spaces: opens (minting if needed) the space keyring.
 * For public spaces: pushes plaintext nodes.
 */
export async function seedSpaceObjectIndex(
  session: Session,
  spaceId: string,
  rooms: SeedRoom[],
  opts?: { visibility?: SpaceVisibility },
): Promise<void> {
  const { encryptor, client } = await getSpaceAccess(spaceId, session, {
    owner: session.userId,
    members: [],
    visibility: opts?.visibility,
  });
  await pushIndexSeed(client, encryptor, spaceId, rooms);
}

/**
 * Headless read-modify-write of a space's unified OBJECT INDEX. Works for both
 * private (encrypt round-trip) and public (plaintext) spaces. Retries up to 3 times
 * on ConflictError.
 */
export async function updateObjectIndex(
  session: Session,
  spaceId: string,
  mutator: (nodes: ObjectNode[], now: number) => ObjectNode[] | null,
  reg?: { owner: string | null; members: string[]; visibility?: SpaceVisibility } | null,
): Promise<void> {
  const { client, encryptor } = await getSpaceAccess(spaceId, session, reg ?? null);
  const pullPath = objIndexPull(spaceId);
  const pushPath = objIndexPush(spaceId);
  const MAX_ATTEMPTS = 3;
  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
    const res = await client.pull(pullPath).catch(() => null);
    const raw = res?.data as Record<string, unknown> | undefined;
    const plain = raw
      ? encryptor
        ? await encryptor.decrypt(raw)
        : raw
      : {};
    const cur = Array.isArray((plain as { objects?: unknown }).objects)
      ? (plain as { objects: ObjectNode[] }).objects
      : [];
    const next = mutator(cur, Date.now());
    if (!next) return;
    const payload = encryptor
      ? await encryptor.encrypt({ objects: next }) as Record<string, unknown>
      : { objects: next };
    try {
      await client.push(pushPath, payload, res?.hash ?? null);
      return;
    } catch (err) {
      if (err instanceof ConflictError && attempt < MAX_ATTEMPTS - 1) continue;
      throw err;
    }
  }
}

File: packages/python/octospaces-sdk/octospaces_sdk/spaces/object_index.py
"""Object index — seed / read / update the ``_index`` doc for a space.

Mirrors ``packages/ts/octospaces-sdk/src/spaces/object-index.ts``.
"""

from __future__ import annotations

import time
from typing import Any, Callable, Optional

from octospaces_sdk.core.types import ObjectNode
from octospaces_sdk.sync.identity import Session
from octospaces_sdk.sync.paths import obj_index_pull, obj_index_push

_MAX_RETRIES = 3


def _strip_invite_node(node: ObjectNode) -> ObjectNode:
    """Strip title/emoji from invite nodes before writing to the shared index."""
    if node.get("access") == "invite":
        stripped = dict(node)
        stripped.pop("title", None)
        stripped.pop("emoji", None)
        return stripped  # type: ignore[return-value]
    return node


async def push_index_seed(client: Any, space_id: str, nodes: list[ObjectNode] | None = None) -> None:
    """Idempotently write an empty (or seeded) object index."""
    path_r = obj_index_pull(space_id)
    path_w = obj_index_push(space_id)

    existing = await client.pull(path_r)
    if existing and existing.get("data"):
        return  # Already seeded

    now = int(time.time() * 1000)
    index = {"v": 2, "objects": nodes or [], "updatedAt": now}
    await client.push(path_w, index)


async def seed_space_object_index(session: Session, space_id: str, nodes: list[ObjectNode] | None = None) -> None:
    await push_index_seed(session.content_client, space_id, nodes)


async def update_object_index(
    session: Session,
    space_id: str,
    mutator: Callable[[list[ObjectNode]], list[ObjectNode]],
    _reg: Any = None,
) -> None:
    """Read-modify-write the object index with up to 3 retries on conflict."""
    path_r = obj_index_pull(space_id)
    path_w = obj_index_push(space_id)

    for _ in range(_MAX_RETRIES):
        doc = await session.content_client.pull(path_r)
        data = doc.get("data", {}) if doc else {}
        base_hash = doc.get("hash") if doc else None
        nodes: list[ObjectNode] = data.get("objects", [])
        now = int(time.time() * 1000)
        updated = mutator(nodes)
        # Strip invite titles before writing
        clean = [_strip_invite_node(n) for n in updated]
        index = {**data, "objects": clean, "updatedAt": now}
        try:
            await session.content_client.push(path_w, index, base_hash=base_hash)
            return
        except Exception as exc:
            if "conflict" in str(exc).lower():
                continue
            raise
    raise RuntimeError("Failed to update object index after retries")


async def read_object_tree(session: Session, space_id: str) -> list[ObjectNode]:
    """Pull the object index and return the flat node list."""
    doc = await session.content_client.pull(obj_index_pull(space_id))
    data = doc.get("data", {}) if doc else {}
    return data.get("objects", [])
